Automattic/jetpack
# Sync: Full Sync: Remove `can_add_to_queue` check in favour of an enqueue "Lock"

Testing Full Sync on very large network sites revealed inconsistencies on the Full Sync enqueue status.
We think that multiple requests could be modifying the enqueue status in parallel resulting in not accurate enqueue status.

Alongside, we were doing a `can_add_to_queue` check late in the enqueue process aborting any enqueuing after having updated the enqueue status.


#### Changes proposed in this Pull Request:
This PR refactors the Full Sync module to add a Lock so two concurrent process can't enqueue at the same time.

In addition it removes multiple checks to `can_add_to_queue` when bulk enqueueing full sync actions.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Bug Fix

#### Testing instructions:
** Test that Full Sync Still works** 
* Apply the Patch
* Initiate a Full Sync.
* `wp jetpack-sync start --blog-id=XXX`
* Check the Jetpack Debug page to see that everything is enqueued as expected.
* Wait for the Sync to complete
* All chunks will have been processed.

#### Proposed changelog entry for your changes:
Bug fixes to Full Sync.

#### Additional Notes
Should add tests to make sure that the "lock" works as expected.
